### PR TITLE
ggml-cuda : perform cublas fp16 matrix multiplication as fp16

### PR DESCRIPTION
**Pull request by slaren**
Improves prompt processing performance with fp16 models.

3090 Ti/WSL2:
| model                          | backend    | ngl | test       |       master t/s |        PR t/s | speedup |
| ------------------------------ | ---------- | --: | ---------- | ---------------: | ------------: | ------: |
| llama 7B mostly F16            | CUDA       |  99 | pp 512     |   1661.19 ± 3.09 | 3984.28 ± 25.45 |    2.40 |

### CODE REVIEW

File: ggml-cuda.cu [hunk 1]
```diff
@@ -6026,33 +6052,78 @@ inline void ggml_cuda_op_mul_mat_cublas(
...
     int ldc = dst->backend == GGML_BACKEND_GPU && id == g_main_device ? ne0 : row_diff;
 
-    CUBLAS_CHECK(cublasSetStream(g_cublas_handles[id], stream));
-    CUBLAS_CHECK(
-        cublasSgemm(g_cublas_handles[id], CUBLAS_OP_T, CUBLAS_OP_N,
-                row_diff, src1_ncols, ne10,
-                &alpha, src0_ddf_i, ne00,
-                        src1_ddf_i,  ne10,
-                &beta,  dst_dd_i,   ldc));
+    // TODO: this may be slower in older architectures with poor fp16 support
```
Comment (Cebtenzzre) [inline ggml-cuda.cu:6062]: Definitely. On my P40, this PR brings FP16 pp512 from 478.92 t/s to only about 10.84 t/s.

**Comment by slaren** [comment]
Is this actually correct? I believe compute capability 7.0 is volta, not turing.

https://github.com/ggerganov/llama.cpp/blob/7d5674dd2d045584993a46102eafa48a31388bdb/ggml-cuda.cu#L82

**Comment by bobqianic** [comment]
> Is this actually correct? I believe compute capability 7.0 is volta, not turing.
> 
> https://github.com/ggerganov/llama.cpp/blob/7d5674dd2d045584993a46102eafa48a31388bdb/ggml-cuda.cu#L82

**The computing capacity of Turing is 7.5, while that of Volta is 7.0. However, Volta also supports FP16.**

image

**Review by ggerganov — APPROVED** [review]

**Comment by ggerganov** [comment]
Hm, this change might actually degrade the TG performance:

Before:

| model                          |       size |     params | backend    | ngl |    threads | test       |              t/s |
| ------------------------------ | ---------: | ---------: | ---------- | --: | ---------: | ---------- | ---------------: |
| LLaMA v2 7B mostly F16         |  12.55 GiB |     6.74 B | CUDA       | 999 |          1 | pp 512     |   3373.39 ± 3.54 |
| LLaMA v2 7B mostly Q8_0        |   6.67 GiB |     6.74 B | CUDA       | 999 |          1 | pp 512     |   2041.33 ± 0.32 |
| LLaMA v2 7B mostly Q4_0        |   3.56 GiB |     6.74 B | CUDA       | 999 |          1 | pp 512     |   2084.74 ± 0.08 |
| LLaMA v2 7B mostly Q4_1        |   3.95 GiB |     6.74 B | CUDA       | 999 |          1 | pp 512     |   2015.38 ± 0.69 |
| LLaMA v2 7B mostly Q5_0        |   4.33 GiB |     6.74 B | CUDA       | 999 |          1 | pp 512     |   2042.62 ± 0.34 |
| LLaMA v2 7B mostly Q5_1        |   4.72 GiB |     6.74 B | CUDA       | 999 |          1 | pp 512     |   1887.36 ± 0.40 |
| LLaMA v2 7B mostly Q6_K        |   5.15 GiB |     6.74 B | CUDA       | 999 |          1 | pp 512     |   2041.73 ± 0.42 |
| LLaMA v2 7B mostly Q5_K - Medium |   4.45 GiB |     6.74 B | CUDA       | 999 |          1 | pp 512     |   1745.14 ± 3.15 |
| LLaMA v2 7B mostly Q5_K - Small |   4.33 GiB |     6.74 B | CUDA       | 999 |          1 | pp 512     |   1674.69 ± 4.07 |
| LLaMA v2 7B mostly Q4_K - Medium |   3.80 GiB |     6.74 B | CUDA       | 999 |          1 | pp 512     |   1816.83 ± 4.45 |
| LLaMA v2 7B mostly Q4_K - Small |   3.59 GiB |     6.74 B | CUDA       | 999 |          1 | pp 512     |   1759.79 ± 3.11 |
| LLaMA v2 7B mostly Q3_K - Medium |   3.07 GiB |     6.74 B | CUDA       | 999 |          1 | pp 512     |   1503.72 ± 1.09 |
| LLaMA v2 7B mostly Q3_K - Small |   2.75 GiB |     6.74 B | CUDA       | 999 |          1 | pp 512     |   1314.71 ± 0.08 |
| LLaMA v2 7B mostly F16         |  12.55 GiB |     6.74 B | CUDA       | 999 |          1 | tg 128     |     72.73 ± 0.02 |
| LLaMA v2 7B mostly Q8_0        |   6.67 GiB |     6.74 B | CUDA       | 999 |          1 | tg 128     |    102.59 ± 0.03 |
| LLaMA v2 7B mostly Q4_0        |   3.56 GiB |     6.74 B | CUDA       | 999 |          1 | tg 128     |    143.87 ± 0.05 |
| LLaMA v2 7B mostly Q4_1        |   3.95 GiB |     6.74 B | CUDA       | 999 |          1 | tg 128     |    142.90 ± 0.03 |
| LLaMA v2 7B mostly Q5_0        |   4.33 GiB |     6.74 B | CUDA       | 999 |          1 | tg 128     |    124.20 ± 0.03 |
| LLaMA v2 7B mostly Q5_1        |   4.72 GiB |     6.74 B | CUDA       | 999 |          1 | tg 128     |    125.25 ± 0.03 |
| LLaMA v2 7B mostly Q6_K        |   5.15 GiB |     6.74 B | CUDA       | 999 |          1 | tg 128     |    103.77 ± 0.01 |
| LLaMA v2 7B mostly Q5_K - Medium |   4.45 GiB |     6.74 B | CUDA       | 999 |          1 | tg 128     |    119.19 ± 0.03 |
| LLaMA v2 7B mostly Q5_K - Small |   4.33 GiB |     6.74 B | CUDA       | 999 |          1 | tg 128     |    122.88 ± 0.04 |
| LLaMA v2 7B mostly Q4_K - Medium |   3.80 GiB |     6.74 B | CUDA       | 999 |          1 | tg 128     |    128.46 ± 0.03 |
| LLaMA v2 7B mostly Q4_K - Small |   3.59 GiB |     6.74 B | CUDA       | 999 |          1 | tg 128     |    133.71 ± 0.02 |
| LLaMA v2 7B mostly Q3_K - Medium |   3.07 GiB |     6.74 B | CUDA       | 999 |          1 | tg 128     |    112.58 ± 0.04 |
| LLaMA v2 7B mostly Q3_K - Small |   2.75 GiB |     6.74 B | CUDA       | 999 |          1 | tg 128     |    100.40 ± 0.07 |

build: 99115f3 (1273)

After:

| model                          |       size |     params | backend    | ngl |    threads | test       |              t/s |
| ------------------------------ | ---------: | ---------: | ---------- | --: | ---------: | ---------- | ---------------: |
| LLaMA v2 7B mostly F16         |  12.55 GiB |     6.74 B | CUDA       | 999 |          1 | pp 512     |  5223.21 ± 11.60 |
| LLaMA v2 7B mostly Q8_0        |   6.67 GiB |     6.74 B | CUDA       | 999 |          1 | pp 512     |   2027.73 ± 1.49 |
| LLaMA v2 7B mostly Q4_0        |   3.56 GiB |     6.74 B | CUDA       | 999 |          1 | pp 512     |   2074.45 ± 1.03 |
| LLaMA v2 7B mostly Q4_1        |   3.95 GiB |     6.74 B | CUDA       | 999 |          1 | pp 512     |   2004.63 ± 0.62 |
| LLaMA v2 7B mostly Q5_0        |   4.33 GiB |     6.74 B | CUDA       | 999 |          1 | pp 512     |   2032.11 ± 0.54 |
| LLaMA v2 7B mostly Q5_1        |   4.72 GiB |     6.74 B | CUDA       | 999 |          1 | pp 512     |   1878.42 ± 0.38 |
| LLaMA v2 7B mostly Q6_K        |   5.15 GiB |     6.74 B | CUDA       | 999 |          1 | pp 512     |   2029.70 ± 1.46 |
| LLaMA v2 7B mostly Q5_K - Medium |   4.45 GiB |     6.74 B | CUDA       | 999 |          1 | pp 512     |   1741.54 ± 2.36 |
| LLaMA v2 7B mostly Q5_K - Small |   4.33 GiB |     6.74 B | CUDA       | 999 |          1 | pp 512     |   1684.32 ± 5.19 |
| LLaMA v2 7B mostly Q4_K - Medium |   3.80 GiB |     6.74 B | CUDA       | 999 |          1 | pp 512     |   1799.69 ± 3.52 |
| LLaMA v2 7B mostly Q4_K - Small |   3.59 GiB |     6.74 B | CUDA       | 999 |          1 | pp 512     |   1749.17 ± 2.88 |
| LLaMA v2 7B mostly Q3_K - Medium |   3.07 GiB |     6.74 B | CUDA       | 999 |          1 | pp 512     |   1497.79 ± 2.87 |
| LLaMA v2 7B mostly Q3_K - Small |   2.75 GiB |     6.74 B | CUDA       | 999 |          1 | pp 512     |   1310.20 ± 0.14 |
| LLaMA v2 7B mostly F16         |  12.55 GiB |     6.74 B | CUDA       | 999 |          1 | tg 128     |     69.43 ± 0.01 |
| LLaMA v2 7B mostly Q8_0        |   6.67 GiB |     6.74 B | CUDA       | 999 |          1 | tg 128     |     91.49 ± 0.02 |
| LLaMA v2 7B mostly Q4_0        |   3.56 GiB |     6.74 B | CUDA       | 999 |          1 | tg 128     |    137.75 ± 0.02 |
| LLaMA v2 7B mostly Q4_1        |   3.95 GiB |     6.74 B | CUDA       | 999 |          1 | tg 128     |    136.58 ± 0.03 |
| LLaMA v2 7B mostly Q5_0        |   4.33 GiB |     6.74 B | CUDA       | 999 |          1 | tg 128     |    112.14 ± 0.01 |
| LLaMA v2 7B mostly Q5_1        |   4.72 GiB |     6.74 B | CUDA       | 999 |          1 | tg 128     |    112.90 ± 0.01 |
| LLaMA v2 7B mostly Q6_K        |   5.15 GiB |     6.74 B | CUDA       | 999 |          1 | tg 128     |     92.56 ± 0.03 |
| LLaMA v2 7B mostly Q5_K - Medium |   4.45 GiB |     6.74 B | CUDA       | 999 |          1 | tg 128     |    102.45 ± 0.01 |
| LLaMA v2 7B mostly Q5_K - Small |   4.33 GiB |     6.74 B | CUDA       | 999 |          1 | tg 128     |    104.18 ± 0.01 |
| LLaMA v2 7B mostly Q4_K - Medium |   3.80 GiB |     6.74 B | CUDA       | 999 |          1 | tg 128     |    111.96 ± 0.02 |
| LLaMA v2 7B mostly Q4_K - Small |   3.59 GiB |     6.74 B | CUDA       | 999 |          1 | tg 128     |    116.10 ± 0.01 |
| LLaMA v2 7B mostly Q3_K - Medium |   3.07 GiB |     6.74 B | CUDA       | 999 |          1 | tg 128     |    103.30 ± 0.04 |
| LLaMA v2 7B mostly Q3_K - Small |   2.75 GiB |     6.74 B | CUDA       | 999 |          1 | tg 128     |     94.84 ± 0.00 |

build: da04003 (1280)

Still testing to verify

**Comment by ggerganov** [comment]
False alarm - forgot to build with `LLAMA_CUDA_MMV_Y=2`

**Comment by whoreson** [comment]
This commit broke llama.cpp on CUDA 10.

identifier "CUBLAS_COMPUTE_16F" is undefined

**Comment by whoreson** [comment]
Let's fix this ok? I can provide SSH access if needed.

**Comment by cebtenzzre** [comment]
> This commit broke llama.cpp on CUDA 10.
> 
> identifier "CUBLAS_COMPUTE_16F" is undefined

Old CUDA versions seem to be a low priority, but you could open a new issue to track this and maybe someone will fix it eventually.

**Comment by ByerRA** [comment]
> This commit broke llama.cpp on CUDA 10.
> 
> identifier "CUBLAS_COMPUTE_16F" is undefined

I am also seeing this as well as a "CUBLAS_TF32_TENSOR_OP_MATH" is undefined error when trying to compile with CUDA 10 and it would be nice to get it fixed or at least a work around we can try to get something working now.

**Pull request merged**